src/core/syscalls/business_logic_syscall_handler.rs

```diff
@@ -0,0 +1,277 @@
+use num_bigint::BigInt;
+use num_traits::ToPrimitive;
+
+//* -----------------------------------
+//* BusinessLogicHandler implementation
+//* -----------------------------------
+
+pub struct BusinessLogicSyscallHandler {
+    tx_execution_context: Rc<RefCell<TransactionExecutionContext>>,
+    events: Rc<RefCell<Vec<OrderedEvent>>>,
+}
+
+impl BusinessLogicSyscallHandler {
+    pub fn new() -> Result<Self, SyscallHandlerError> {
+        let events = Rc::new(RefCell::new(Vec::new()));
+        let tx_execution_context = Rc::new(RefCell::new(TransactionExecutionContext::new()));
+        Ok(BusinessLogicSyscallHandler {
+            events,
+            tx_execution_context,
+        })
+    }
+}
+
+impl SyscallHandler for BusinessLogicSyscallHandler {
+    // trait functions
+    fn emit_event(
+        &self,
+        vm: &VirtualMachine,
+        syscall_ptr: Relocatable,
+    ) -> Result<(), SyscallHandlerError> {
+        let SyscallRequest::EmitEvent(request) =
+            self._read_and_validate_syscall_request("emit_event", vm, syscall_ptr)?;
+
+        let keys_len = bigint_to_usize(&request.keys_len)?;
+        let data_len = bigint_to_usize(&request.data_len)?;
+
+        let order = self.tx_execution_context.borrow_mut().n_emitted_events;
+        let keys: Vec<BigInt> = get_integer_range(vm, &request.keys, keys_len)?;
+        let data: Vec<BigInt> = get_integer_range(vm, &request.data, data_len)?;
+
+        self.events
+            .borrow_mut()
+            .push(OrderedEvent::new(order, keys, data));
+
+        // Update events count.
+        self.tx_execution_context.borrow_mut().n_emitted_events += 1;
+        Ok(())
+    }
+
+    fn send_message_to_l1(&self, _vm: VirtualMachine, _syscall_ptr: Relocatable) {
+        todo!()
+    }
+
+    fn _get_tx_info_ptr(&self, _vm: VirtualMachine) {
+        todo!()
+    }
+    fn _deploy(&self, _vm: VirtualMachine, _syscall_ptr: Relocatable) -> i32 {
+        todo!()
+    }
+
+    fn _read_and_validate_syscall_request(
+        &self,
+        syscall_name: &str,
+        vm: &VirtualMachine,
+        syscall_ptr: Relocatable,
+    ) -> Result<SyscallRequest, SyscallHandlerError> {
+        self.read_syscall_request(syscall_name, vm, syscall_ptr)
+    }
+
+    fn _call_contract(
+        &self,
+        _syscall_name: &str,
+        _vm: VirtualMachine,
+        _syscall_ptr: Relocatable,
+    ) -> Vec<i32> {
+        todo!()
+    }
+    fn _get_caller_address(&self, _vm: VirtualMachine, _syscall_ptr: Relocatable) -> i32 {
+        todo!()
+    }
+    fn _get_contract_address(&self, _vm: VirtualMachine, _syscall_ptr: Relocatable) -> i32 {
+        todo!()
+    }
+    fn _storage_read(&self, _address: i32) -> i32 {
+        todo!()
+    }
+    fn _storage_write(&self, _address: i32, _value: i32) {
+        todo!()
+    }
+    fn _allocate_segment(&self, _vm: VirtualMachine, _data: Vec<MaybeRelocatable>) -> Relocatable {
+        todo!()
+    }
+}
+
+//* -------------------
```

Comment:
This function looks useful enough to move it to a utils module to reuse it easily

Reply:
this also applies for get_integer_range

Reply:
done